logical/framework/openapi.go

```diff
@@ -0,0 +1,548 @@
+
+import (
+	"fmt"
+	"regexp"
+	"sort"
+	"strings"
+
+	log "github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/vault/helper/wrapping"
+	"github.com/hashicorp/vault/logical"
+	"github.com/hashicorp/vault/version"
+	"github.com/mitchellh/mapstructure"
+)
+
+// OpenAPI specification (OAS): https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md
+const OASVersion = "3.0.2"
+
+// NewOASDocument returns an empty OpenAPI document.
+func NewOASDocument() *OASDocument {
+	return &OASDocument{
+		Version: OASVersion,
+		Info: oasInfo{
+			Title:       "HashiCorp Vault API",
+			Description: "HTTP API that gives you full access to Vault. All API routes are prefixed with `/v1/`.",
+			Version:     version.GetVersion().Version,
+			License: oasLicense{
+				Name: "Mozilla Public License 2.0",
+				URL:  "https://www.mozilla.org/en-US/MPL/2.0",
+			},
+		},
+		Paths: make(map[string]*oasPathItem),
+	}
+}
+
+type OASDocument struct {
+	Version string                  `json:"openapi"`
+	Info    oasInfo                 `json:"info"`
+	Paths   map[string]*oasPathItem `json:"paths"`
+}
+
+type oasInfo struct {
+	Title       string     `json:"title"`
+	Description string     `json:"description"`
+	Version     string     `json:"version"`
+	License     oasLicense `json:"license"`
+}
+
+type oasLicense struct {
+	Name string `json:"name"`
+	URL  string `json:"url"`
+}
+
+type oasPathItem struct {
+	Description     string         `json:"description,omitempty"`
+	Parameters      []oasParameter `json:"parameters,omitempty"`
+	Sudo            bool           `json:"x-vault-sudo,omitempty" mapstructure:"x-vault-sudo"`
+	Unauthenticated bool           `json:"x-vault-unauthenticated,omitempty" mapstructure:"x-vault-unauthenticated"`
+	Create          bool           `json:"x-vault-create,omitempty" mapstructure:"x-vault-create"`
+
+	Get    *OASOperation `json:"get,omitempty"`
+	Post   *OASOperation `json:"post,omitempty"`
+	Delete *OASOperation `json:"delete,omitempty"`
+}
+
+// NewOASOperation creates an empty OpenAPI Operations object.
+func NewOASOperation() *OASOperation {
+	return &OASOperation{
+		Responses: make(map[string]*oasResponse),
+	}
+}
+
+type OASOperation struct {
+	Summary     string                  `json:"summary,omitempty"`
+	Description string                  `json:"description,omitempty"`
+	Tags        []string                `json:"tags,omitempty"`
+	Parameters  []oasParameter          `json:"parameters,omitempty"`
+	RequestBody *oasRequestBody         `json:"requestBody,omitempty"`
+	Responses   map[string]*oasResponse `json:"responses"`
+	Deprecated  bool                    `json:"deprecated,omitempty"`
+}
+
+type oasParameter struct {
+	Name        string     `json:"name"`
+	Description string     `json:"description,omitempty"`
+	In          string     `json:"in"`
+	Schema      *oasSchema `json:"schema,omitempty"`
+	Required    bool       `json:"required,omitempty"`
+	Deprecated  bool       `json:"deprecated,omitempty"`
```

Comment:
Do Description and Deprecated need to be duplicated between here and the values in `Schema`? (Would embedding `*oasSchema` be an even better approach?)

Reply:
They do serve different roles.

Schemas are a base definition type used in multiple places (parameters, requests bodies, response bodies, even recursively), and have their own descriptions and deprecation info. For example our request data is one schema, with each property being its own schema that corresponds to a field and has a description, deprecation state, and so forth.

Parameters (usually query) have a name and reference a schema. The description/deprecation here applies to parameter itself. We tend to think of parameters as being a string, but they could technically be an encoded object with a more complicated schema.

Personally I find the layout quite convoluted, but all of these structures _very_ closely mirror OpenAPI, which allows us to JSON-encode directly. The "input" side (stuff in `path.go`) is where we can optimize the definition experience to allow flexibility without having tons of unneeded openapi structure. The processing bridges the two.

Reply:
Thanks for the info!